### ISSUE TITLE
CI/CD: Script to run CI/CD tests on incoming email patches

### ISSUE THREAD

**Issue by danieldaquino** [damus-io/damus#1719]
**Acceptance criteria:**
- Create a script that:
    - Listens to patches@damus.io and applies the patch to a new git branch
- Configure Xcode Cloud to listen to those branches and trigger a test run

**Security considerations:**
- Given that anyone can send emails to patches@damus.io, we need to be careful about where that code goes.
    - Should those new branches be added to the main damus repo? Or in some mirror/fork that is less visible (to prevent other people from accidentally pulling potentially bad branches)?
    - Should we implement email address whitelisting, where only known senders can have their patch automatically trigger a CI/CD build?

**Comment by alltheseas** [comment]
Is this the same as https://github.com/damus-io/damus/issues/1711 ? I can close 1711 if needed

**Comment by danieldaquino** [comment]
@alltheseas, yes! They should be the same task. Sorry for opening a duplicate, we can merge the two issues into one

**Comment by jb55** [comment]
On Wed, Nov 15, 2023 at 09:56:11AM -0800, Daniel D’Aquino wrote:
> **Acceptance criteria:**
> - Create a script that:
>     - Listens to ***@***.*** and applies the patch to a new git branch
> - Configure Xcode Cloud to listen to those branches and trigger a test run

xcode cloud is already configured to build any new branch

>**Security considerations:**

> - Given that anyone can send emails to ***@***.***, we need to be
> careful about where that code goes.

What is the worst a patch could do? There are no secrets in the repo.
This is also true for github actions and new PRs.

> - Should those new branches be added to the main damus repo? Or in
> some mirror/fork that is less visible (to prevent other people from
> accidentally pulling potentially bad branches)?

Pulling bad branches would be pretty rare, most people only interact
with master. We can just prefix the branches with patchset- or something to
make it clear they are ci builds. The script would need some way to
build canonical branch names. Maybe we could just use the first 8 chars
of the hash of the message-id

> - Should we implement email address whitelisting, where only known
> senders can have their patch automatically trigger a CI/CD build?

Definitely can look into that if there are any issues. If someone is
clever enough to put together an email patch to try to be malicious then
good for them.

**Comment by jb55** [comment]
On Wed, Nov 15, 2023 at 10:50:20AM -0800, William Casarin wrote:
>On Wed, Nov 15, 2023 at 09:56:11AM -0800, Daniel D’Aquino wrote:
>>- Should those new branches be added to the main damus repo? Or in
>>some mirror/fork that is less visible (to prevent other people from
>>accidentally pulling potentially bad branches)?
>
>Pulling bad branches would be pretty rare, most people only interact
>with master. We can just prefix the branches with patchset- or something to
>make it clear they are ci builds. The script would need some way to
>build canonical branch names. Maybe we could just use the first 8 chars
>of the hash of the message-id

Now that I think of it, it may be better to have more descriptive branch
names. Maybe the first 4 words of the subject snake-cased.
